src/main/java/de/tum/in/www1/artemis/domain/Exercise.java

```diff
@@ -597,6 +595,69 @@ public void filterForCourseDashboard(List<StudentParticipation> participations,
         }
     }
 
+    /**
+     * Filter for appropriate submission. Relevance in the following order:
+     * - submission with rated result
+     * - submission with unrated result (late submission)
+     * - no submission with any result > latest submission
+     *
+     * @param submissions that need to be filtered
+     * @return filtered submission
+     */
+    private Submission findAppropriateSubmission(Set<Submission> submissions) {
+        if (submissions == null || submissions.isEmpty())
+            return null;
+
+        List<Submission> submissionsWithRatedResult = new ArrayList<>();
+        List<Submission> submissionsWithUnratedResult = new ArrayList<>();
+        List<Submission> submissionsWithoutResult = new ArrayList<>();
+
+        for (Submission submission : submissions) {
+            Result result = submission.getResult();
+            if (result != null) {
+                if (result.isRated()) {
+                    submissionsWithRatedResult.add(submission);
+                }
+                else {
+                    submissionsWithUnratedResult.add(submission);
+                }
+            }
+            else {
+                submissionsWithoutResult.add(submission);
+            }
+        }
+
+        if (submissionsWithRatedResult.size() > 0) {
+            return getLatestSubmission(submissionsWithRatedResult);
+        }
+        else if (submissionsWithUnratedResult.size() > 0) {
+            return getLatestSubmission(submissionsWithUnratedResult);
+        }
+        else {
+            return getLatestSubmission(submissionsWithoutResult);
+        }
+    }
+
+    /**
+     * Sorts a list of submission by submissionDate and returns the latest one
+     *
+     * @param submissions list of submissions that should be sorted
+     * @return submission
+     */
+    private Submission getLatestSubmission(List<Submission> submissions) {
```

Comment:
This is a very complicated way to get the latest entry in a list.

E.g. in lambda:

```
Submission latestSubmission = list.stream().max(Comparator.comparing(s -> s.getSubmissionDate())).ifPresent(...);
```

Reply:
Thank you for bringing this to my attention! I like the functional stuff :-)